### PR TITLE
Enhance CSVExtension source to work in Linux style docker 

### PR DESCRIPTION
**Pull request by sagizagagi**

### CODE REVIEW

File: CSVSource/README.md [hunk 1]
```diff
@@ -16,6 +16,8 @@ In this way, in cases where the extension is not running during file updates, it
 In addition to handling delimiter-separated fields, this connector can extract fields from fixed positions in the record.
 This is configured as described below.
 
+the extension source support docker, when running in docker is detected, the windows file path is changed to UNIX file path, so the same configuration can be used for docker and non-docker activate.
```
Comment (namirfa) [inline CSVSource/README.md:19]: Change to:

`When running this connector as a docker image, the file path is dynamically converted to a UNIX file path. This means that the same configuration can be used for both docker and non-docker runtimes.`

**Review by namirfa — CHANGES_REQUESTED** [review]
Some high level comments about cleaning up the code and fixing REAMDE.

After that is done, re-request a review and we will take a look at the code changes once more.

File: CSVSource/src/main/java/io/vantiq/extsrc/CSVSource/CSV.java [hunk 2]
```diff
@@ -241,25 +256,49 @@ public void setupCSV(ExtensionWebSocketClient oClient, String fileFolderPath, St
...
+            TimerTask task = new TimerTask() {
+                @Override
+                public void run() {
+                    log.info("TimerTask Start working on existing file in folder {}", tmpFileFolderPath);
+                    handleExistingFiles(tmpFileFolderPath);
+
+                }
+            };
+            // Create new Timer, and schedule the task according to the pollTime
+            inDockerTime = new Timer("executePolling");
```
Comment (fhcarter) [inline CSVSource/src/main/java/io/vantiq/extsrc/CSVSource/CSV.java:280]: This doesn't really seem to have anything to do with Docker at all. This variable should just be `timerTask` or something, shouldn't it?

File: CSVSource/src/main/java/io/vantiq/extsrc/CSVSource/CSV.java [hunk 3]
```diff
@@ -547,49 +597,30 @@ void handleExistingFiles(String fileFolderPath) {
...
-                            Path newPath = ((WatchEvent<Path>) watchEvent).context();
-                            log.debug("Ignored path modified: {}", newPath.toString());
-                        }
-                    }
-                    key.reset();
-                }
-            } catch (InterruptedException ex1) {
-                log.error("processThread inloop failure", ex1);
-            }
+    String fixFileFolderPathForDocker(String filePath) {
```
Comment (fhcarter) [inline CSVSource/src/main/java/io/vantiq/extsrc/CSVSource/CSV.java:616]:
```suggestion
    String fixFileFolderPathForUnix(String filePath) {
```

File: CSVSource/README.md (hunk 1 above)
Comment (sagizagagi) [reply CSVSource/README.md:19]: done

File: CSVSource/README.md [hunk 4]
```diff
@@ -215,6 +218,7 @@ the configuration should be similar to
 * **processExistingFiles**: Optional. If set to `true`, the service will process all files already existing in the folder `fileFolderPath` (filtered using `filePrefix` and the `fileExtension`). Otherwise the service will process only new files.  Default is`false`.
 * **extensionAfterProcessing**: Optional. Rename the file after it has been processed to avoid reprocessing (_e.g._ for cases where `processExistingFiles` is set to `true`).  The default value is combination of the 'fileExtension' and `done`.  For example `.csv.done` when `fileExtension` set to `.csv`.
 * **deleteAfterProcessing**: Optional. Delete the processed file only if processed successfully to avoid reprocessing in cases where `processExistingFiles` is set to `true`. Default value is `false`.
+* **pollTime**: Optional, Default is 30000 Milliseconds, the frequency the extension checks the target folder for new files to process.
```
[the remark this answers is not in the document]
Comment (sagizagagi) [reply CSVSource/README.md:221]: done

File: CSVSource/src/main/java/io/vantiq/extsrc/CSVSource/CSV.java [hunk 5]
```diff
@@ -133,21 +137,26 @@
     Map<String, Object> options;
 
     // Control of thread should continue to work, set to false in Stop.
-    boolean bContinue = true;
+    // boolean bContinue = true;
```
[the remark this answers is not in the document]
Comment (sagizagagi) [reply CSVSource/src/main/java/io/vantiq/extsrc/CSVSource/CSV.java:140]: done

File: CSVSource/src/main/java/io/vantiq/extsrc/CSVSource/CSV.java [hunk 6]
```diff
@@ -133,21 +137,26 @@
...
 
+    boolean isRunnigInDocker = isRunningInsideDocker();
     // Components used
-    WatchService serviceWatcher;
+    // WatchService serviceWatcher;
```
[the remark this answers is not in the document]
Comment (sagizagagi) [reply CSVSource/src/main/java/io/vantiq/extsrc/CSVSource/CSV.java:144]: done

File: CSVSource/src/main/java/io/vantiq/extsrc/CSVSource/CSV.java [hunk 7]
```diff
@@ -201,6 +210,12 @@ void prepareConfigurationData() {
             deleteAfterProcessing = (boolean) options.get("deleteAfterProcessing");
         }
 
+        if (options.get("pollTime") != null) {
+            pollTime = (Integer) options.get("pollTime");
+        } else {
+            pollTime = 30000;
```
[the remark this answers is not in the document]
Comment (sagizagagi) [reply CSVSource/src/main/java/io/vantiq/extsrc/CSVSource/CSV.java:216]: done

File: CSVSource/src/main/java/io/vantiq/extsrc/CSVSource/CSV.java [hunk 8]
```diff
@@ -241,25 +256,49 @@ public void setupCSV(ExtensionWebSocketClient oClient, String fileFolderPath, St
...
+             * 
+             * // working on files already that exist in folder. if
+             * (options.get("processExistingFiles") != null) { Object processExistingFiles =
+             * options.get("processExistingFiles"); if (processExistingFiles instanceof
+             * Boolean) if ((boolean) processExistingFiles) {
+             * log.info("Start working on existing file in folder {}", fileFolderPath);
+             * handleExistingFiles(fileFolderPath); } }
+             * 
+             * }
+             */
```
[the remark this answers is not in the document]
Comment (sagizagagi) [reply CSVSource/src/main/java/io/vantiq/extsrc/CSVSource/CSV.java:300]: done

File: CSVSource/src/main/java/io/vantiq/extsrc/CSVSource/CSV.java [hunk 9]
```diff
@@ -447,7 +493,7 @@ void handleExistingFiles(String fileFolderPath) {
                 file.createNewFile();
 
                 try (FileOutputStream fos = new FileOutputStream(file);
-                    BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(fos))) {
+                        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(fos))) {
```
[the remark this answers is not in the document]
Comment (sagizagagi) [reply CSVSource/src/main/java/io/vantiq/extsrc/CSVSource/CSV.java:496]: no, this is how the formator is doing,

File: CSVSource/src/main/java/io/vantiq/extsrc/CSVSource/CSV.java [hunk 10]
```diff
@@ -547,49 +597,30 @@ void handleExistingFiles(String fileFolderPath) {
         }
     }
 
+    public static Boolean isRunningInsideDocker() {
```
[the remark this answers is not in the document]
Comment (sagizagagi) [reply CSVSource/src/main/java/io/vantiq/extsrc/CSVSource/CSV.java:600]: done

**Review by fhcarter — CHANGES_REQUESTED** [review]
typos.

Also, your docker-stuff isn't really doing docker.  It's checking for isUnix.  Please rename.

Finally, your claim that it does configurations cross system isn't quite right.  It can convert windows -> unix, but not the other way around...

File: extjsdk/src/main/java/io/vantiq/extjsdk/ExtensionWebSocketClient.java [hunk 11]
```diff
@@ -298,7 +298,7 @@ public void sendNotification(Object data) {
      * receipt of a response message.
      */
     void acknowledgeNotification() {
-        outstandingNotifications.release();
+            outstandingNotifications.release();
```
[the remark this answers is not in the document]
Comment (sagizagagi) [reply extjsdk/src/main/java/io/vantiq/extjsdk/ExtensionWebSocketClient.java:301]: I have added the fix I need if it is OK with you. done

[the remark this answers is not in the document]
Comment (fhcarter) [reply extjsdk/src/main/java/io/vantiq/extjsdk/ExtensionWebSocketClient.java:301]: all you changed here is spacing.  This doesn't change any behavior

If there's something to be fixed here, it needs, as I said, to be in a separate PR.

**Comment by sagizagagi** [comment]
That because I didn’t roll back completely , I tried to do what you ask and filled an issue about that.

File: CSVSource/src/main/java/io/vantiq/extsrc/CSVSource/CSV.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (fhcarter) [reply CSVSource/src/main/java/io/vantiq/extsrc/CSVSource/CSV.java:472]: tell formatter to not use tabs and to indent by 4 spaces.

File: extjsdk/src/main/java/io/vantiq/extjsdk/ExtensionWebSocketClient.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (fhcarter) [reply extjsdk/src/main/java/io/vantiq/extjsdk/ExtensionWebSocketClient.java]: this needs to be in a separate PR.  As noted a few times.

And, don't reformat the whole file to add 1 if statement.

File: HikVisionSource/src/main/java/io/vantiq/extsrc/HikVisionSource/HikVision.java [hunk 12]
```diff
@@ -0,0 +1,1467 @@
+/*
```
Comment (fhcarter) [inline HikVisionSource/src/main/java/io/vantiq/extsrc/HikVisionSource/HikVision.java:1]: Remove this file.

**Comment by sagizagagi** [comment]
I don’t think the other way around is necessary. Atlist not now. 

And there might be other activities that one needs to enhance in that situation.

File: CSVSource/README.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (sagizagagi) [reply CSVSource/README.md]: when the system runs on a windows machine, no conversion is done to the file system, 
it is Vantiq recommendation to use docker with a Linux ( and not Microsoft ) as such when it runs in a container, it will always behave like Linux

**Comment by namirfa** [comment]
@sagizagagi

Looks like you checked in a lot of formatting changes to the ExtensionWebSocketClient class, probably a result of your IDE's formatting. Those changes should all be undone so that the class remains identical to the current version in master.

**Comment by sagizagagi** [comment]
I tried to rollback, hope I managed to do so.

File: extjsdk/src/main/java/io/vantiq/extjsdk/ExtensionWebSocketClient.java [hunk 13]
```diff
@@ -275,14 +275,17 @@ public void sendNotification(Object data) {
         ExtensionServiceMessage msg = new ExtensionServiceMessage("");
         msg.fromMap(m);
         if (isConnected()) {
+            Semaphore localOutstandingNotifications = outstandingNotifications;
```
Comment (fhcarter) [inline extjsdk/src/main/java/io/vantiq/extjsdk/ExtensionWebSocketClient.java:278]: you should just revert these changes completely.  Namir has fixed the problem as it was slightly more involved.  In the end, your PR should have no changes to this file.

File: CSVSource/src/main/java/io/vantiq/extsrc/CSVSource/CSV.java (hunk 2 above)
Comment (sagizagagi) [reply CSVSource/src/main/java/io/vantiq/extsrc/CSVSource/CSV.java:275]: yes, a leftover evaluation of the learning curve, I will fix it. done

File: CSVSource/src/main/java/io/vantiq/extsrc/CSVSource/CSV.java [hunk 14]
```diff
@@ -133,21 +137,26 @@
     Map<String, Object> options;
 
     // Control of thread should continue to work, set to false in Stop.
-    boolean bContinue = true;
+    // boolean bContinue = true;
 
+    boolean isRunnigInDocker = isRunningInsideDocker();
```
[the remark this answers is not in the document]
Comment (sagizagagi) [reply CSVSource/src/main/java/io/vantiq/extsrc/CSVSource/CSV.java:139]: done

File: CSVSource/src/main/java/io/vantiq/extsrc/CSVSource/CSV.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (sagizagagi) [reply CSVSource/src/main/java/io/vantiq/extsrc/CSVSource/CSV.java:472]: done

[the remark this answers is not in the document]
Comment (sagizagagi) [reply CSVSource/src/main/java/io/vantiq/extsrc/CSVSource/CSV.java:472]: done

File: CSVSource/README.md [hunk 15]
```diff
@@ -109,7 +111,8 @@ The Configuration document may look similar to the following example:
             "maxQueuedTasks": 4,
             "processExistingFiles": true,
             "extensionAfterProcessing": "csv.done",
-            "deleteAfterProcessing": false
+            "deleteAfterProcessing": false,
+            "pollTime":30000
```
[the remark this answers is not in the document]
Comment (sagizagagi) [reply CSVSource/README.md:115]: done

File: CSVSource/src/main/java/io/vantiq/extsrc/CSVSource/CSV.java (hunk 3 above)
Comment (sagizagagi) [reply CSVSource/src/main/java/io/vantiq/extsrc/CSVSource/CSV.java:596]: done

**Comment by sagizagagi** [comment]
I failed to revert, so I just got the code from master ….

Namir's enhancement was already there….

File: CSVSource/src/main/java/io/vantiq/extsrc/CSVSource/CSV.java [hunk 16]
```diff
@@ -547,49 +597,30 @@ void handleExistingFiles(String fileFolderPath) {
         }
     }
 
+    public static Boolean isRunningInsideDocker() {
+
+        try (Stream<String> stream = Files.lines(Paths.get("/proc/1/cgroup"))) {
```
[the remark this answers is not in the document]
Comment (fhcarter) [reply CSVSource/src/main/java/io/vantiq/extsrc/CSVSource/CSV.java:581]: You aren't addressing this.  You need to.

File: CSVSource/README.md [hunk 17]
```diff
@@ -16,6 +16,8 @@ In this way, in cases where the extension is not running during file updates, it
 In addition to handling delimiter-separated fields, this connector can extract fields from fixed positions in the record.
 This is configured as described below.
 
+When running this connector as a docker image, the file path is dynamically converted to a UNIX file path. This means that the same configuration can be used for both docker and non-docker runtimes.
```
Comment (fhcarter) [inline CSVSource/README.md:19]: This is simply not true.  If the configuration is a windows one, it'll be converted to unix.  The other way is not true.

File: CSVSource/src/main/java/io/vantiq/extsrc/CSVSource/CSV.java [hunk 18]
```diff
@@ -377,6 +391,9 @@ void handleExistingFiles(String fileFolderPath) {
             Map<String, Object> body = (Map<String, Object>) request.get(BODY_KEYWORD);
             checkedAttribute = PATH_KEYWORD;
             pathStr = (String) body.get(PATH_KEYWORD);
+            if (isRunningInDocker) {
```
Comment (fhcarter) [inline CSVSource/src/main/java/io/vantiq/extsrc/CSVSource/CSV.java:394]: again, you don't really care about running in docker -- you care about things being unix or windows style file paths.  Please make _that_ clear.

File: CSVSource/src/main/java/io/vantiq/extsrc/CSVSource/CSV.java (hunk 16 above)
[the remark this answers is not in the document]
Comment (fhcarter) [reply CSVSource/src/main/java/io/vantiq/extsrc/CSVSource/CSV.java:581]: If you want a simpler fix that has the same effect, check 

```
if (File.separator == '/') {
   // then it's unix...
}

**Review by fhcarter — CHANGES_REQUESTED** [review]
`ExtensionWebSocketClient.java` is still in the list.  It shouldn't be/

Need to address text & names & README text about the docker stuff. You don't (or shouldn't) care if you are looking at docker or now.  You care about if it's a unix file path.

File: CSVSource/src/main/java/io/vantiq/extsrc/CSVSource/CSV.java (hunk 18 above)
Comment (sagizagagi) [reply CSVSource/src/main/java/io/vantiq/extsrc/CSVSource/CSV.java:394]: done, change to isRunningInLinux

File: CSVSource/README.md (hunk 17 above)
Comment (sagizagagi) [reply CSVSource/README.md:19]: done

File: CSVSource/src/main/java/io/vantiq/extsrc/CSVSource/CSV.java (hunk 16 above)
[the remark this answers is not in the document]
Comment (sagizagagi) [reply CSVSource/src/main/java/io/vantiq/extsrc/CSVSource/CSV.java:581]: done

File: extjsdk/src/main/java/io/vantiq/extjsdk/ExtensionWebSocketClient.java (hunk 13 above)
Comment (sagizagagi) [reply extjsdk/src/main/java/io/vantiq/extjsdk/ExtensionWebSocketClient.java:278]: I have spent more than an hour already trying to revert. and yes, I already pulled the correction of Namir from master.
just forgive me on that one, any way it will merge with the master as it is a similar content.

File: CSVSource/README.md [hunk 19]
```diff
@@ -16,6 +16,8 @@ In this way, in cases where the extension is not running during file updates, it
 In addition to handling delimiter-separated fields, this connector can extract fields from fixed positions in the record.
 This is configured as described below.
 
+When running this connector on Linux and the supplied file path are windoes style, the file path is dynamically converted to a UNIX file path. This means that the same configuration can be used for both windows and non-windows runtimes.
```
Comment (fhcarter) [inline CSVSource/README.md:19]:
```suggestion
When running this connector on Linux and the supplied file path are windows style, the file path is dynamically converted to a UNIX file path. This means that the same configuration can be used for both windows and non-windows runtimes if and only if the configuration is done in windows path style.
```

**Review by fhcarter — APPROVED** [review]
You still need to correct the description in the README to accurately reflect what it does -- windows -> unix but not the other way around.

Also, the fix to the web socket client was trivial -- like I said, checkout master, pull master, checkout your branch, merge  master.  Problem goes away -- I did it.

**Review by namirfa — APPROVED** [review]
LGTM - pending fixes for Fred's comments.
